### PR TITLE
Enhance logic for section links

### PR DESCRIPTION
**Pull request by nealedj**
### What is the context of this PR?
Delegates the logic for working how whether a section is not started, started, complete or skipped to a separate module. This is used by the navigation builder to work out which block to link to when a user clicks on a section link. It's also used to work out if a questionnaire has been completed and can be submitted.

- If a section has not been started the user should link to the first block.
- If it has been started then the user will link to the first incomplete block.
- If it has been completed and the final block is a summary or confirmation block then the user will link to the final block
- If it has been completed and the final block is not a summary or confirmation block then the user will link to the first block

### How to review 
- Test section navigation using surveys that make use of routing, skip conditions and repeated groups

### CODE REVIEW

File: app/questionnaire/completeness.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        incomplete_locations = (
+            location for location, is_complete in
+            self._get_block_states_for_group(
+                group, group_instance=group_instance)
+            if not is_complete
+        )
+        return next(incomplete_locations, None)
+
+    def is_block_valid_on_path(self, location, block):
+        if block['type'] == 'Question':
```
Comment (jonnyshaw89) [inline app/questionnaire/completeness.py]: if type isn't `Question` this will return `None`?
Should it not return `false`?

File: app/questionnaire/completeness.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            return location in self.routing_path
+
+    def _get_block_states_for_group(self, group, group_instance=None):
+        if group_instance is not None:
+            current_instance = max_instance = group_instance
+        else:
+            max_instance = self._calculate_no_of_repeats(group) - 1
+            current_instance = 0
+
+        while current_instance <= max_instance:
```
Comment (jonnyshaw89) [inline app/questionnaire/completeness.py]: Does this need to be a `while` loop?
we have all the information upfront so it could just be a `for`?

File: app/questionnaire/completeness.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                location = Location(group['id'], current_instance, block['id'])
+
+                if not self.is_block_valid_on_path(location, block):
+                    continue
+
+                yield location, self.is_block_complete(location)
+
+            current_instance += 1
+
+    def _calculate_no_of_repeats(self, group):
```
Comment (jonnyshaw89) [inline app/questionnaire/completeness.py]: This logic is identical to that on `PathFinder` Can't this be reused?

File: app/questionnaire/navigation.py [hunk 4]
```diff
@@ -34,36 +38,40 @@ def build_navigation(self, current_group_id, current_group_instance):
...
 
-            if first_block_id in summary_or_confirmation_blocks \
-                    and self._can_reach_summary_confirmation(first_group_id, first_block_id) is False:
+            if is_summary_or_confirmation_section and \
+                    not self._can_reach_summary_confirmation(section):
                 continue
 
-            repeating_rule = self.schema.get_repeat_rule(
-                self.schema.get_group(first_group_id))
+            # assume that a repeating group will always be the first group in a section
```
Comment (jonnyshaw89) [inline app/questionnaire/navigation.py]: This assumption is incorrect for `census_household`
https://github.com/ONSdigital/eq-survey-runner/blob/33d2b28556cb2cd4c3cb20eae178e78a95bd7d72/data/en/census_household.json#L3662

File: data/en/1_0005.json [hunk 5]
```diff
@@ -362,6 +362,20 @@
                                             "voluntary employee contributions to, and annual profit from, profit related pay schemes"
                                         ]
                                     },
+                                    {
```
Comment (jonnyshaw89) [inline data/en/1_0005.json]: Is this change related?

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: data/en/1_0005.json (hunk 5 above)
Comment (nealedj) [reply data/en/1_0005.json]: No - looks like a bad merge. Will fix.

File: app/questionnaire/completeness.py (hunk 2 above)
Comment (nealedj) [reply app/questionnaire/completeness.py]: Not sure - I think might be clearer as a `while` loop in this instance. Doing this as a for loop would be:

`for current_instance in range(start_instance, max_instance + 1):`

That feels less readable to me.

File: app/questionnaire/completeness.py (hunk 1 above)
Comment (nealedj) [reply app/questionnaire/completeness.py]: Done

**Comment by jonnyshaw89** [comment]
@nealedj the PR description says "If is has been completed then the user will link to the final block." I have tested this on `test_navigation_completeness.json` and it takes me to the last block in the section

**Comment by jonnyshaw89** [comment]
When completing `census_household.json` the visitors section appears before I have been asked about visitors. when i click on the visitors link i get redirected to the first incomplete block of the survey and not the 1st block of the visitors section

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]
A few observations above

**Review by dwyeradam — CHANGES_REQUESTED** [review]
I tried the e-commerce survey as that's got a lot of routing/skip conditions. When you select No to the initial question you get routed around the whole of the rest of the survey straight to the comments section at the end. However when this happens the first Section title goes from the Navigation on the right.

**Comment by nealedj** [comment]
@jonnyshaw89 RE:

> @nealedj the PR description says "If is has been completed then the user will link to the final block." I have tested this on test_navigation_completeness.json and it takes me to the last block in the section

That wasn't quite right - I've updated the description now.

**Comment by jonnyshaw89** [comment]
@nealedj in `census_household.json` the visitors section is still visible when it shouldn't be.

**Comment by jonnyshaw89** [comment]
The pre submission completeness check still checks all block types. if i navigate to another group at the end of a section, when i go to submission page, i am redirected back the the interstitial that was not submitted. only Question pages should be required to be submitted

File: tests/app/questionnaire/test_completeness.py [hunk 6]
```diff
@@ -8,6 +8,16 @@
 
 
 class TestCompleteness(AppContextTestCase): # pylint: disable=too-many-public-methods
+    def test_no_groups(self):
```
Comment (jonnyshaw89) [inline tests/app/questionnaire/test_completeness.py]: `test_section_with_no_groups_not_marked_as_complete_when_.....`??

File: app/helpers/completeness_helper.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+from flask_login import current_user, login_required
+from werkzeug.local import LocalProxy
+
+from app.globals import get_answer_store, get_metadata, get_completed_blocks
+from app.helpers.path_finder_helper import path_finder
+from app.questionnaire.completeness import Completeness
+
+
+@login_required
+def get_completeness():
```
Comment (jonnyshaw89) [inline app/helpers/completeness_helper.py]: I think this should live in `globals.py` as that is the place we currently house this sort of logic. 
The "helpers" are currently more utility methods or wrappers

Comment (jonnyshaw89) [reply app/helpers/completeness_helper.py]: This could even be made as a wrapper eg `path_finder_helper.full_routing_path_required`??

File: app/questionnaire/navigation.py [hunk 8]
```diff
@@ -22,6 +22,9 @@ def __init__(self, schema, answer_store, metadata, completed_blocks, routing_pat
         self.completed_blocks = completed_blocks
         self.routing_path = routing_path
 
+        self.completeness = Completeness(
```
Comment (jonnyshaw89) [inline app/questionnaire/navigation.py]: Why not use your new get_completeness() method?
I think its off that we have a method that constructs this for us yet we don't consistently use it?
To quote 'O-Town' I think we should have it "All or nothing at all"

File: app/questionnaire/navigation.py [hunk 9]
```diff
@@ -34,147 +37,128 @@ def build_navigation(self, current_group_id, current_group_instance):
...
-                    and self._can_reach_summary_confirmation(first_group_id, first_block_id) is False:
+            if is_summary_or_confirmation_section and \
+                    not self._can_reach_summary_confirmation(section):
                 continue
 
-            repeating_rule = self.schema.get_repeat_rule(
-                self.schema.get_group(first_group_id))
+            # if the first group in a section is a repeating group then repeat the section
+            # navigation link for each repeat instead of rendering the section title
+            repeating_rule = self.schema.get_repeat_rule(non_skipped_groups[0])
```
Comment (jonnyshaw89) [inline app/questionnaire/navigation.py]: I think we should also change this to sections that contain one 1 group that repeats.

File: app/questionnaire/questionnaire_schema.py [hunk 10]
```diff
@@ -130,6 +130,42 @@ def get_questions_for_block(block_json):
         for question_json in block_json.get('questions', []):
             yield question_json
 
+    @staticmethod
+    def is_summary_section(section):
+        """Summary section is a section with a single Summary group
+        """
+        if section['groups']:
```
Comment (jonnyshaw89) [inline app/questionnaire/questionnaire_schema.py]: If we change validator to require at least 1 group for a section then we don't need this check.

File: app/questionnaire/questionnaire_schema.py [hunk 11]
```diff
@@ -130,6 +130,42 @@ def get_questions_for_block(block_json):
...
+            return QuestionnaireSchema.is_summary_group(section['groups'][0])
```
Comment (jonnyshaw89) [inline app/questionnaire/questionnaire_schema.py]: I think we should check all groups in the section, as we don't currently restrict a summary to be in its own section. I don't think.

File: app/questionnaire/questionnaire_schema.py [hunk 12]
```diff
@@ -130,6 +130,42 @@ def get_questions_for_block(block_json):
...
+
+        return False
+
+    @staticmethod
+    def is_summary_group(group):
+        """Summary group is a group with a single block of type "Summary"
+        """
+        if group['blocks']:
+            return group['blocks'][0]['type'] == 'Summary'
```
Comment (jonnyshaw89) [inline app/questionnaire/questionnaire_schema.py]: Same 2 comments as above. we shouldn't have a group with no blocks and we don't stop summary being in a group with other blocks so we should check all the blocks

File: app/questionnaire/questionnaire_schema.py [hunk 13]
```diff
@@ -130,6 +130,42 @@ def get_questions_for_block(block_json):
...
+
+        return False
+
+    @staticmethod
+    def is_confirmation_section(section):
+        """Confirmation section is a section with a single confirmation group
+        """
+        if section['groups']:
+            return QuestionnaireSchema.is_confirmation_group(section['groups'][0])
```
Comment (jonnyshaw89) [inline app/questionnaire/questionnaire_schema.py]: same again

File: app/questionnaire/questionnaire_schema.py [hunk 14]
```diff
@@ -130,6 +130,42 @@ def get_questions_for_block(block_json):
...
+
+        return False
+
+    @staticmethod
+    def is_confirmation_group(group):
+        """Confirmation group is a group with a single block of type "Confirmation"
+        """
+        if group['blocks']:
+            return group['blocks'][0]['type'] == 'Confirmation'
```
Comment (jonnyshaw89) [inline app/questionnaire/questionnaire_schema.py]: and again

File: app/views/questionnaire.py [hunk 15]
```diff
@@ -285,9 +287,9 @@ def get_view_submission(eq_id, form_type):  # pylint: disable=unused-argument
     return redirect(url_for('post_submission.get_thank_you', eq_id=eq_id, form_type=form_type))
 
 
-def _redirect_to_latest_location(routing_path, collection_id, eq_id, form_type):
-    latest_location = path_finder.get_latest_location(get_completed_blocks(current_user),
-                                                      routing_path=routing_path)
+def _redirect_to_latest_location(collection_id, eq_id, form_type):
+    latest_location = completeness.get_latest_location()
```
Comment (jonnyshaw89) [inline app/views/questionnaire.py]: I think if we make `get_completeness` a wrapper then it would be more obvious where `completeness` is coming from.

File: app/views/questionnaire.py [hunk 16]
```diff
@@ -306,7 +308,8 @@ def _render_page(full_routing_path, block, current_location, post_form=None):
 
 
 def _is_valid_location(routing_path, location):
-    return location in routing_path
+    is_survey_complete = not completeness.get_first_incomplete_location_in_survey()
+    return location in routing_path or is_survey_complete
```
Comment (jonnyshaw89) [inline app/views/questionnaire.py]: Im not sure how this helps?

File: data/en/census_household.json [hunk 17]
```diff
@@ -3668,7 +3668,7 @@
                             "when": [{
                                 "id": "overnight-visitors-answer",
                                 "condition": "equals",
-                                "value": "0"
+                                "value": 0
```
Comment (jonnyshaw89) [inline data/en/census_household.json:3669]: 👍

File: data/en/test_0203.json [hunk 18]
```diff
@@ -8,334 +8,346 @@
...
-                                }
-                            ],
-                            "description": "If possible, this should be for the period {{exercise.start_date|format_date}} to {{exercise.end_date|format_date}}.",
-                            "id": "reporting-period-question",
-                            "title": "What are the dates of the sales period you are reporting for?",
-                            "type": "DateRange"
-                        },
-                        {
-                            "guidance": {
+            "id": "default-section",
```
Comment (jonnyshaw89) [inline data/en/test_0203.json]: Why have we changed this schema?

File: data/en/test_0205.json [hunk 19]
```diff
@@ -8,361 +8,372 @@
...
-                                    "q_code": "12",
-                                    "label": "Period to",
-                                    "type": "Date",
-                                    "mandatory": true
-                                }
-                            ]
-                        },
-                        {
-                            "guidance": {
+            "id": "default-section",
```
Comment (jonnyshaw89) [inline data/en/test_0205.json]: and this one?

File: tests/app/questionnaire/test_completeness.py [hunk 20]
```diff
@@ -0,0 +1,483 @@
...
+            Location('coffee-group', 0, 'coffee'),
+            Location('coffee-group', 0, 'response-yes'),
+        ]
+        progress = Completeness(
+            schema, AnswerStore(), completed_blocks, routing_path, metadata={})
+
+        progress_value = progress.get_state_for_section('coffee-section')
+        self.assertEqual(Completeness.COMPLETED, progress_value)
+
+    def test_all_sections_complete_none(self):
```
Comment (jonnyshaw89) [inline tests/app/questionnaire/test_completeness.py]: The test name doesn't make sense on its own, until you read the subsequent 2 tests

File: tests/app/questionnaire/test_completeness.py [hunk 21]
```diff
@@ -0,0 +1,483 @@
...
+            Location('coffee-group', 0, 'coffee'),
+            Location('coffee-group', 0, 'response-yes'),
+            Location('toast-group', 0, 'toast'),
+        ]
+
+        progress = Completeness(
+            schema, AnswerStore(), completed_blocks, routing_path, metadata={})
+        self.assertTrue(progress.all_sections_complete())
+
+    def test_any_section_complete_none(self):
```
Comment (jonnyshaw89) [inline tests/app/questionnaire/test_completeness.py:115]: same comment about the name

File: tests/app/questionnaire/test_completeness.py [hunk 22]
```diff
@@ -0,0 +1,483 @@
...
+            Location('coffee-group', 0, 'coffee'),
+            Location('coffee-group', 0, 'response-yes'),
+            Location('toast-group', 0, 'toast'),
+        ]
+
+        progress = Completeness(
+            schema, AnswerStore(), completed_blocks, routing_path, metadata={})
+        self.assertFalse(progress.get_first_incomplete_location_in_survey())
+
+    def get_first_incomplete_location_in_survey_completed_skipped_interstitials(self):
```
Comment (jonnyshaw89) [inline tests/app/questionnaire/test_completeness.py]: doesn't this test also test everything tested in the test above?

File: tests/app/questionnaire/test_completeness.py [hunk 23]
```diff
@@ -0,0 +1,483 @@
...
+            Location('payment-details', 0, 'security-code'),
+            Location('payment-details', 0, 'security-code-interstitial'),
+            Location('extra-cover-items-group', 0, 'extra-cover-items'),
+            Location('confirmation-group', 0, 'confirmation'),
+        ]
+        progress = Completeness(
+            schema, AnswerStore(), completed_blocks, routing_path, metadata={})
+        self.assertFalse(progress.get_first_incomplete_location_in_survey())
+
+    def test_get_latest_location_no_completed_blocks(self):
```
Comment (jonnyshaw89) [inline tests/app/questionnaire/test_completeness.py:477]: I wonder if we should be specifying what we expect on these tests?
`test_get_latest_location_no_completed_blocks_returns_first_location`

File: tests/app/questionnaire/test_questionnaire_schema.py [hunk 24]
```diff
@@ -204,3 +204,67 @@ def test_group_has_questions_returns_false_when_group_doesnt_have_questionnaire_
...
+            }]
+        }
+
+        schema = QuestionnaireSchema(survey_json)
+        self.assertTrue(schema.is_confirmation_section(schema.get_section('section-1')))
+        self.assertTrue(schema.is_confirmation_group(schema.get_group('group-1')))
+        self.assertFalse(schema.is_summary_section(schema.get_section('section-1')))
+        self.assertFalse(schema.is_summary_group(schema.get_group('group-1')))
+
+    def test_is_summary_confirmation_no_groups(self):
```
Comment (jonnyshaw89) [inline tests/app/questionnaire/test_questionnaire_schema.py]: We shouldn't be able to have this scenario

File: tests/app/questionnaire/test_questionnaire_schema.py [hunk 25]
```diff
@@ -204,3 +204,67 @@ def test_group_has_questions_returns_false_when_group_doesnt_have_questionnaire_
...
+        survey_json = {'sections': [{'id': 'section-1', 'groups': []}]}
+        schema = QuestionnaireSchema(survey_json)
+        self.assertFalse(schema.is_confirmation_section(schema.get_section('section-1')))
+        self.assertFalse(schema.is_summary_section(schema.get_section('section-1')))
+
+    def test_is_summary_confirmation_no_blocks(self):
```
Comment (jonnyshaw89) [inline tests/app/questionnaire/test_questionnaire_schema.py]: We shouldn't be able to have this scenario

**Review by jonnyshaw89 — DISMISSED** [review]

File: app/helpers/completeness_helper.py (hunk 7 above)
Comment (nealedj) [reply app/helpers/completeness_helper.py]: Probably makes more sense as a function in `globals`

File: app/questionnaire/navigation.py (hunk 8 above)
Comment (nealedj) [reply app/questionnaire/navigation.py]: Ha - yep good point. I missed this when putting in that function.

**Review by dwyeradam — CHANGES_REQUESTED** [review]
I have just tried the census household schema. Upon completing the visitors section and pressing continue I get routed back to the first uncompleted block which is correct. However the highlighting on Section navigation isn't working when this happens.

File: app/questionnaire/questionnaire_schema.py (hunk 11 above)
Comment (nealedj) [reply app/questionnaire/questionnaire_schema.py]: @jonnyshaw89 Given that we now have a section summary. Perhaps we should add this restriction? I think I put in this assumption for that reason.

File: app/questionnaire/questionnaire_schema.py (hunk 10 above)
Comment (nealedj) [reply app/questionnaire/questionnaire_schema.py]: Good point 👍

File: app/questionnaire/questionnaire_schema.py (hunk 11 above)
Comment (jonnyshaw89) [reply app/questionnaire/questionnaire_schema.py]: rename it to `is_final_summary_section` and check the block type is `Summary` not `SectionSummary`?

File: app/views/questionnaire.py (hunk 15 above)
Comment (nealedj) [reply app/views/questionnaire.py]: I did that for a reason... but I can't remember what it was now. I'll find out and clarify.

File: app/views/questionnaire.py (hunk 16 above)
Comment (nealedj) [reply app/views/questionnaire.py]: I did that for a reason... but I can't remember what it was now. I'll find out and clarify.

File: tests/app/questionnaire/test_completeness.py (hunk 22 above)
Comment (nealedj) [reply tests/app/questionnaire/test_completeness.py]: Yes, true.

File: tests/app/questionnaire/test_completeness.py (hunk 23 above)
Comment (nealedj) [reply tests/app/questionnaire/test_completeness.py:477]: I find that can result in really long method names. Prefer that the method name describes the scenario and the assertion should indicate the expected outcome.

Thoughts? @jonnyshaw89

Comment (jonnyshaw89) [reply tests/app/questionnaire/test_completeness.py:477]: Yeh, i'm not die hard either way.
Our code is scattered with examples of both.

**Comment by nealedj** [comment]
@MIslam92 @shahi645 @MebinAbraham @davecharles 

I've added you all as requested reviewers. Please focus primarily on functionally testing the application with these changes rather than on the code changes.

File: app/questionnaire/navigation.py [hunk 26]
```diff
@@ -34,147 +42,131 @@ def build_navigation(self, current_group_id, current_group_instance):
         if navigation_block is None or navigation_block.get('visible', True) is False:
             return None
 
-        summary_or_confirmation_blocks = self.schema.get_summary_and_confirmation_blocks()
-
         navigation = []
 
         for section in self.schema.sections:
             non_skipped_groups = self._get_non_skipped_groups(section)
             if not non_skipped_groups:
```
Comment (dwyeradam) [inline app/questionnaire/navigation.py:48]: Why aren't we using the Completeness class for this method? At the moment Navigation still seems to be evaluating skip conditions when that has already been done by the Completeness class. Ideally completeness would be the one stop shop for defining what state a section, group or block is in.

**Review by dwyeradam — CHANGES_REQUESTED** [review]

**Review by davecharles — COMMENTED** [review]
I've tested your changes with the following schemas and have no observations:
- census_household.json
- test_skip_condition.json
- test_skip_condition_block.json
- test_skip_condition_group.json
- test_routing_on_multiple_select.json
- test_routing_date_*.json

Also checked with labour_force.json and behaviour is as expected.

**Review by MIslam92 — COMMENTED** [review]
I've tested your changes with the following schemas and seems to behave the way as stated in the PR/Card:

- `1_0005.json` (Including the Introduction in the Nav however does not get marked as complete when you start the survey probs due to its type being 'Introduction'?)
- `census_household.json`
- `e_commerce.json`
- `test_navigation.json`
- `test_navigation_*.json`
- `test_section_summary.json`

**Comment by MebinAbraham** [comment]
@nealedj Can you update your branch, I'm getting behaviours that should no longer happen (ie new PRs that went in since you last updated).

I've merged locally. Will update if I find anything.

**Review by MebinAbraham — CHANGES_REQUESTED** [review]
This PR #1530 seems to cause issues with your section completeness.
Tested schema: `1_0005.json`
If you try to visit a block that should no longer be in the routing path, it currently lets me visit that block and when I try to continue it will throw a 500 error.

If that's not very clear, you can recreate the error following this:
1. Launch `1_0005.json`
2. On `Pay Pattern` section choose 2 options, ie `Weekly` and `Fortnightly`
3. Complete the `Weekly Pay` and `General Comments` section.
4. Go to `Fortnightly Pay` section and copy the URL.
5. Go to `Pay Pattern` section and uncheck `Fortnightly Pay` section and click Continue.
6. Now visit the `Fortnightly` section using the URL.

No. 6 above should take you to 'Summary' since all sections are complete and not the section that shouldn't be in the routing path. If you also click `Continue` after No. 6, it will throw an error 500.

It is working as expected on master. Sorry for making it long-winded.

**Comment by MIslam92** [comment]
@MebinAbraham I just tested your scenario and don't get a 500 error and can submit as normal?

**Comment by MIslam92** [comment]
@MebinAbraham I just re ran your scenario seems to work fine for me and no 500's?

File: app/questionnaire/navigation.py (hunk 26 above)
Comment (nealedj) [reply app/questionnaire/navigation.py:48]: Have addressed this now.

**Comment by nealedj** [comment]
Good catch @MebinAbraham. I recreated that and fix it in fb5cf509fc7222ba7029cdb485f5990aaea284cc

File: app/questionnaire/navigation.py [hunk 27]
```diff
@@ -34,147 +41,115 @@ def build_navigation(self, current_group_id, current_group_instance):
...
 
     def _get_non_skipped_groups(self, section):
         return [
-            group['id'] for group in self._get_visible_groups_for_section(section)
-            if not self._should_skip_group(group)
+            group for group in section['groups']
+            if self.completeness.get_state_for_group(group) != Completeness.SKIPPED
         ]
 
     def _should_skip_group(self, group):
```
Comment (dwyeradam) [inline app/questionnaire/navigation.py]: `_should_skip_group` method is no longer needed. Which will also help your code coverage.

**Review by dwyeradam — CHANGES_REQUESTED** [review]

File: app/questionnaire/navigation.py (hunk 27 above)
Comment (nealedj) [reply app/questionnaire/navigation.py]: Oh I thought I'd deleted that - ta!

File: app/globals.py [hunk 28]
```diff
@@ -67,6 +68,23 @@ def get_completed_blocks(user):
...
+
+    completeness_object = g.get('_completeness')
+
+    if completeness_object is None:
+        metadata = get_metadata(user)
+        answer_store = get_answer_store(user)
+        completed_blocks = get_completed_blocks(user)
+        routing_path = path_finder.get_full_routing_path()
+
+        completeness_object = g.completeness = Completeness(
```
Comment (shahi645) [inline app/globals.py]: should be `_completeness`, after your explanation yesterday?

File: app/questionnaire/completeness.py [hunk 29]
```diff
@@ -0,0 +1,188 @@
...
+
+        def eval_state(state_to_compare):
+            return (state == state_to_compare for state in block_states)
+
+        group_state = self.NOT_STARTED
+
+        if all(eval_state(self.SKIPPED)):
+            group_state = self.SKIPPED
+
+        elif not self.routing_path:
```
Comment (shahi645) [inline app/questionnaire/completeness.py:97]: A lot of duplication here. Not very familiar with methods within methods. Can these be simplified into the one method?

**Review by shahi645 — CHANGES_REQUESTED** [review]

**Comment by shahi645** [comment]
@nealedj Also flake 8 failing on lint script for an unused import

File: app/globals.py (hunk 28 above)
Comment (nealedj) [reply app/globals.py]: Oh yeah - well remembered!

File: app/questionnaire/completeness.py (hunk 29 above)
Comment (nealedj) [reply app/questionnaire/completeness.py:97]: In this case, not that I can think of at the moment. Although the checks look similar the logic for deriving a group's state from its blocks states is different to the logic that derive a section's state from its groups.

**Comment by nealedj** [comment]
@MebinAbraham @dwyeradam @shahi645 ready for re-review

File: tests/functional/spec/navigation.spec.js [hunk 30]
```diff
@@ -200,7 +200,6 @@ describe('Navigation', function() {
     .click(SecurityCodePage.submit())
     .click(SkipInterstitialPage.no())
     .click(SkipInterstitialPage.submit())
-    .then(helpers.isSectionComplete('Payment Details')).should.eventually.be.false
```
Comment (dwyeradam) [inline tests/functional/spec/navigation.spec.js:203]: Really you should have changed this to true as your code changes the behaviour here so Interstitial pages no longer count towards section completeness. Not a blocker though but if any further changes are made it might be worth changing this.

**Review by MebinAbraham — COMMENTED** [review]
Quick question. If a section has 2 questions and the first question is mandatory. Should you be able to view the Section Summary (SS) without completing all the questions in that section?

Cause at the moment, I am able to view the SS without answering the mandatory question. If I click `Continue` on the SS I would go to the next Section and once that is complete it would route me back to the first mandatory question in the first section.

Just wondering if this is normal or after answering the 2nd question in the first section, I should be routed back to the first mandatory question I bypassed?

**Comment by nealedj** [comment]
@MebinAbraham good call. No, that shouldn't happen. Although it wasn't introduced as part of this PR. It's reasonably minor because the main summary still won't allow the questionnaire to be submitted if there are incomplete blocks.

I'll raise a separate card to address this.

File: app/questionnaire/completeness.py (hunk 29 above)
Comment (dwyeradam) [reply app/questionnaire/completeness.py:97]: @nealedj, are you sure that we can't reduce the duplication around eval_state. That looks the same for group and block. @shahi645 be strong and force your point.

Comment (shahi645) [reply app/questionnaire/completeness.py:97]: It would save a good few lines of code 😉

Comment (dwyeradam) [reply app/questionnaire/completeness.py:97]: Actually @shahi645 I led us up a garden path. David is correct. There are two subtle differences that I missed. Groups also uses routing path to evaluate NOT_STARTED and an extra step for checking COMPLETED. Lines 97/98, and 103/104 are group only checks.

Comment (shahi645) [reply app/questionnaire/completeness.py:97]: Thats fine! I thought originally we might be able to, was worth asking the question

**Review by dwyeradam — APPROVED** [review]
This does conflict slightly with return to Summary after editing a question. In this PR section completeness doesn't rely on SectionSummary and Interstitial pages being complete. However routing back to summary does. Ignoring interstitial in routing back to summary would be an issue if the last block is an interstitial as it would never be seen as you'd skip it straight to summary. Although I don't like tech debt I think this will need more thought and may be better as a PR in it's own right. Current working only gives a bad user experience in an edge case.

**Review by MebinAbraham — APPROVED** [review]
Apart from the issues identified that can be raised as separate cards. 
Looks good to me from a functional perspective.
👍

File: app/questionnaire/navigation.py [hunk 31]
```diff
@@ -200,9 +162,58 @@ def _generate_link_names(self, label_answer_ids):
...
+    def _get_location_for_group(self, group, group_instance=None):
+        """Gets the location for a link used for a repeating group
+        """
+        group_completeness = self.completeness.get_state_for_group(group, group_instance=group_instance)
+        first_incomplete = self.completeness.get_first_incomplete_location_in_group(group, group_instance=group_instance)
+
+        return self._get_location_for_completeness(
+            group_completeness,
+            group,
+            group,
```
Comment (shahi645) [inline app/questionnaire/navigation.py:188]: Should this be reapeated_group or something?

**Comment by shahi645** [comment]
@nealedj LGTM apart from that little comment i made but otherwise 👍

File: app/questionnaire/navigation.py (hunk 31 above)
Comment (nealedj) [reply app/questionnaire/navigation.py:188]: You mean because the same value is passed twice? No, in this case that's intentional. This method gets a link for a repeating group, all of the possible locations are within the same group so the value passed for `first_group` and `last_group` are both the same.

Comment (shahi645) [reply app/questionnaire/navigation.py:188]: Ahhh ok yes that was it, in which case this is G2G

**Review by shahi645 — APPROVED** [review]
